java/fory-format/src/main/java/org/apache/fory/format/row/binary/CompactBinaryRow.java

```diff
@@ -0,0 +1,157 @@
+    super(schema);
+    this.fixedOffsets = fixedOffsets;
+    bitmapOffset = fixedOffsets[fixedOffsets.length - 1];
+    allFieldsNotNullable = CompactBinaryRowWriter.allNotNullable(schema.getFields());
+  }
+
+  @Override
+  protected int computeBitmapWidthInBytes() {
+    // cannot use field due to initialization order
+    if (CompactBinaryRowWriter.allNotNullable(schema.getFields())) {
+      return 0;
+    }
+    return CompactBinaryRowWriter.headerBytes(schema);
+  }
+
+  @Override
+  public boolean isNullAt(final int ordinal) {
+    if (allFieldsNotNullable) {
+      return false;
+    }
+    return super.isNullAt(ordinal);
+  }
+
+  // TODO: this should use StableValue once it's available
+  @Override
+  public int getOffset(final int ordinal) {
+    return baseOffset + fixedOffsets[ordinal];
+  }
+
+  @Override
+  public MemoryBuffer getBuffer(final int ordinal) {
+    final int fixedWidthBinary = CompactBinaryRowWriter.fixedWidthFor(schema, ordinal);
+    if (fixedWidthBinary >= 0) {
+      if (isNullAt(ordinal)) {
+        return null;
+      }
+      return getBuffer().slice(getOffset(ordinal), fixedWidthBinary);
+    } else {
+      return super.getBuffer(ordinal);
+    }
+  }
+
+  @Override
+  public byte[] getBinary(final int ordinal) {
+    final int fixedWidthBinary = CompactBinaryRowWriter.fixedWidthFor(schema, ordinal);
+    if (fixedWidthBinary >= 0) {
+      if (isNullAt(ordinal)) {
+        return null;
+      }
+      final byte[] bytes = new byte[fixedWidthBinary];
+      getBuffer().get(getOffset(ordinal), bytes, 0, fixedWidthBinary);
+      return bytes;
+    } else {
+      return super.getBinary(ordinal);
+    }
+  }
+
+  @Override
+  protected BinaryRow getStruct(final int ordinal, final Field field, final int extDataSlot) {
+    if (isNullAt(ordinal)) {
+      return null;
+    }
+    final int fixedWidthBinary = CompactBinaryRowWriter.fixedWidthFor(schema, ordinal);
+    if (fixedWidthBinary == -1) {
+      return super.getStruct(ordinal, field, extDataSlot);
+    }
+    if (extData[extDataSlot] == null) {
+      extData[extDataSlot] = DataTypes.createSchema(field);
+    }
+    final BinaryRow row = newRow((Schema) extData[extDataSlot]);
+    row.pointTo(getBuffer().slice(getOffset(ordinal), fixedWidthBinary), 0, fixedWidthBinary);
+    return row;
+  }
+
+  @Override
+  protected Schema newSchema(final Field field) {
+    return CompactBinaryRowWriter.sortSchema(super.newSchema(field));
+  }
+
+  @Override
+  protected BinaryRow newRow(final Schema schema) {
+    // TODO: avoid re-computing these offsets
+    return new CompactBinaryRow(schema, CompactBinaryRowWriter.fixedOffsets(schema));
+  }
+
+  @Override
+  protected BinaryArray newArray(final Field field) {
+    return new CompactBinaryArray(field);
+  }
+
+  @Override
+  protected BinaryMap newMap(final Field field) {
+    return new CompactBinaryMap(field);
+  }
+
+  @Override
+  protected int isNullBitmapOffset() {
```

Comment:
this looks like a method return `boolean`, how abouting naming it as `nullabilityBitmapOffset`